### PR TITLE
Fix get_messages endpoint and add NoDataError and use it in request methods

### PR DESCRIPTION
**Pull request by AndrewLester**

### CODE REVIEW

File: schoolopy/main.py [hunk 1]
```diff
@@ -31,12 +31,12 @@ def _get(self, path):
...
         """
+        response = self.schoology_auth.oauth.get(url='%s%s?limit=%s' % (self.api_host, path, self.limit), headers=self.schoology_auth._request_header(), auth=self.schoology_auth.oauth.auth)
+        response.raise_for_status()
         try:
-            response = self.schoology_auth.oauth.get(url='%s%s?limit=%s' % (self.api_host, path, self.limit), headers=self.schoology_auth._request_header(), auth=self.schoology_auth.oauth.auth)
-            response.raise_for_status()
             return response.json()
         except JSONDecodeError:
-            return {}
+            raise NoDataError(f'Get request to {response.url} failed with response {response.text}')
```
Comment (ErikBoesen) [inline schoolopy/main.py:39]: I would recommend making the `response.text` a little more separated from the rest of the error message. For example:
```py
raise NoDataError(f'Get request to {response.url} failed: {response.text}')
```

File: schoolopy/main.py [hunk 2]
```diff
@@ -46,12 +46,12 @@ def _post(self, path, data):
...
         """
+        response = self.schoology_auth.oauth.post(url='%s%s?limit=%s' % (self.api_host, path, self.limit), json=data, headers=self.schoology_auth._request_header(), auth=self.schoology_auth.oauth.auth)
+        response.raise_for_status()
         try:
-            response = self.schoology_auth.oauth.post(url='%s%s?limit=%s' % (self.api_host, path, self.limit), json=data, headers=self.schoology_auth._request_header(), auth=self.schoology_auth.oauth.auth)
-            response.raise_for_status()
             return response.json()
         except json.decoder.JSONDecodeError:
-            return {}
+            raise NoDataError(f'Post request to {response.url} failed with response {response.text}')
```
Comment (ErikBoesen) [inline schoolopy/main.py:54]: Same thing here

File: schoolopy/main.py [hunk 3]
```diff
@@ -61,20 +61,22 @@ def _put(self, path, data):
...
         """
+        response = self.schoology_auth.oauth.put(url='%s%s?limit=%s' % (self.api_host, path, self.limit), json=data, headers=self.schoology_auth._request_header(), auth=self.schoology_auth.oauth.auth)
+        response.raise_for_status()
         try:
-            response = self.schoology_auth.oauth.put(url='%s%s?limit=%s' % (self.api_host, path, self.limit), json=data, headers=self.schoology_auth._request_header(), auth=self.schoology_auth.oauth.auth)
-            response.raise_for_status()
             return response.json()
         except json.decoder.JSONDecodeError:
-            return {}
+            raise NoDataError(f'Put request to {response.url} failed with response {response.text}')
```
Comment (ErikBoesen) [inline schoolopy/main.py:69]: And here

File: schoolopy/main.py [hunk 4]
```diff
@@ -1925,8 +1927,16 @@ def get_roles(self):
     def get_role(self, role_id):
         return Role(self._get('roles/%s' % role_id))
 
-    def get_messages(self):
-        return [MessageThread(raw) for raw in self._get('messages')['message']]
+    def get_messages(self, message_folder):
+        """
+        List messages in one of the valid message folders.
+
+        :param message_folder: Either 'sent' or 'inbox'
```
Comment (ErikBoesen) [inline schoolopy/main.py:1934]: I almost feel like this is worth creating an enum... not quite though.

**Review by ErikBoesen — CHANGES_REQUESTED** [review]

File: schoolopy/main.py (hunk 4 above)
Comment (AndrewLester) [reply schoolopy/main.py:1934]: Originally I had a type annotation that only allowed 'sent' or 'inbox'. But I decided that having just one type annotation in the whole file would be weird.

**Review by erikboesen-powerschool — APPROVED** [review]

**Comment by erikboesen-powerschool** [comment]
Oops wrong account

**Review by ErikBoesen — APPROVED** [review]

**Pull request merged**